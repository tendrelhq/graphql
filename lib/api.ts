import type { Connection } from "@/schema/system/pagination";
import { assert, type RawPaginationArgs, assertNonNull, map } from "./util";

type PaginationMeta = Omit<Connection<void>, "edges">;

export function extractPageInfo(res: Response): PaginationMeta {
  const contentRange = assertNonNull(
    res.headers.get("Content-Range"),
    "cannot paginate without content-range",
  );

  const [start, end, count] = map(contentRange.split("/"), cr =>
    [...cr[0].split("-"), cr[1]].map(Number),
  );
  assert(Number.isFinite(start));
  assert(Number.isFinite(end));
  assert(Number.isFinite(count));

  if (res.status === 206) {
    return {
      pageInfo: {
        hasNextPage: true,
        hasPreviousPage: false,
        endCursor: end.toString(),
        startCursor: start.toString(),
      },
      totalCount: count,
    };
  }

  return {
    pageInfo: {
      hasNextPage: false,
      hasPreviousPage: false,
    },
    totalCount: count,
  };
}

export function constructHeadersFromArgs(
  args: RawPaginationArgs,
  preferences?: string[],
): Headers {
  const headers = new Headers({
    Prefer: ["count=estimated"].concat(preferences ?? []).join(","),
  });
  const lower = Number(args.after ?? 0);
  const count = Math.min(args.first ?? 100, 100);
  const upper = lower + count - 1;
  headers.append("Range", `${lower}-${upper}`);
  return headers;
}
