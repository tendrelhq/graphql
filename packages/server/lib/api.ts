import type { Connection } from "@/schema/system/pagination";
import { assert, type RawPaginationArgs, assertNonNull, map } from "./util";

type PaginationMeta = Omit<Connection<void>, "edges">;

export function extractPageInfo(res: Response): PaginationMeta {
  const contentRange = assertNonNull(
    res.headers.get("Content-Range"),
    "cannot paginate without content-range",
  );

  const range = map(contentRange.split("/"), cr => {
    const [start, end] = cr[0].split("-");
    return {
      start: start,
      end: end,
      count: Number(cr[1]),
    };
  });
  assert(Number.isFinite(range.count), `invalid count: ${contentRange}`);

  if (range.start === "*" || res.status !== 206) {
    // Range not satisfiable
    return {
      pageInfo: {
        hasNextPage: false,
        hasPreviousPage: false,
      },
      totalCount: range.count,
    };
  }

  const start = Number(range.start);
  const end = Number(range.end);
  assert(Number.isFinite(start), `invalid start: ${contentRange}`);
  assert(Number.isFinite(end), `invalid end: ${contentRange}`);
  return {
    pageInfo: {
      hasNextPage: end < range.count - 1,
      hasPreviousPage: start > 0,
      endCursor: end.toString(),
      startCursor: start.toString(),
    },
    totalCount: range.count,
  };
}

export function constructHeadersFromArgs(
  args: RawPaginationArgs,
  preferences?: Record<string, string> & {
    count?: "estimated" | "exact" | "planned";
  },
): Headers {
  const prefers = preferences
    ? Object.keys(preferences)
        .map(key => `${key}=${preferences[key]}`)
        .join(",")
    : "count=estimated";
  const headers = new Headers({ Prefer: prefers });
  const lower = Number(args.after ?? 0);
  const count = Math.min(args.first ?? 100, 100);
  const upper = lower + count - 1;
  headers.append("Range-Unit", "items");
  headers.append("Range", `${lower}-${upper}`);
  return headers;
}
