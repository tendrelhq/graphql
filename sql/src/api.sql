CREATE SCHEMA api;
COMMENT ON SCHEMA api IS '
# Tendrel REST API

## Authentication and authorization

Most of the public api requires two tokens to be present in every request:

1. An application token, in JWT format via the X-Tendrel-App header.
2. An authorization token, in JWT format via the Authorization header.

The application token is essentially an OAuth 2.0 client secret, where the
"application" (e.g. a mobile app) is the OAuth 2.0 client. Application tokens
are required for all api calls.

The authorization token is a Tendrel issued JWT that uniquely identifies the
user making the request. Most api calls require a valid authorization token.
A client may exchange an IDP-issued security token for a Tendrel-issued token by
calling the /token api, which is implemented as an [OAuth 2.0 Token
Exchange](https://datatracker.ietf.org/doc/html/rfc8693).

### The "anonymous" role

The anonymous (`anon`) role can only be used to perform a select few operations
that do not require an authorization token. In particular, the `anon` role can
be used to signup for a Tendrel account as well as in various "introspection"
related operations, e.g. introspecting the OpenAPI schema. This role is the
default role for HTTP requests which do not include an authorization token.

## Localization

Tendrel is not a content management system. It does, however, provide mechanisms
that allow for content to be dynamically localized according to user preference.

Dynamic localization can be customized using various HTTP headers:

- `Accept-Language: en` specifies the locale.
- `Prefer: timezone=America/Denver` specifies the timezone.
';

