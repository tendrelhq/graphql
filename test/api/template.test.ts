import { describe, expect, test } from "bun:test";
import { baseurl } from "@/test/api/constants";
import { pg } from "../prelude";

describe("/api/template", () => {
  test("create", async () => {
    const response = await pg
      .from("entity_template")
      .insert({
        owner: "f90d618d-5de7-4126-8c65-0afb700c6c61",
        name: "My first template!",
      })
      .select("name")
      .rollback();
    expect(response).toMatchObject({
      data: expect.arrayContaining([
        {
          name: "My first template!",
        },
      ]),
      status: 201,
    });
  });

  test.skip("read", async () => {
    const url = new URL(
      "/template?select=id,display_name(value),fields(id,display_name(value),type(id,display_name(value)))&order=id",
      baseurl,
    );
    const response = await fetch(url, {
      method: "GET",
      headers: {
        Range: "0-19",
      },
    });
    expect(response.json()).resolves.toMatchSnapshot();
  });

  test.skip("update", async () => {
    const template = "not-a-real-uuid"; // FIXME: need a demo customer
    const url = new URL(
      `/template?id=eq.${template}&select=id,_order`,
      baseurl,
    );
    const response = await fetch(url, {
      method: "PATCH",
      headers: {
        Prefer: "return=representation",
      },
      body: JSON.stringify({ _order: 2 }),
    });
    expect(response.json()).resolves.toMatchSnapshot();
  });

  test("delete", async () => {
    const owner = "f90d618d-5de7-4126-8c65-0afb700c6c61";
    const template = "0b9f3142-e7ed-4f78-8504-ccd2eb505075";
    const response = await pg
      .rpc("delete_entity_template", { owner, id: template })
      .select("id,_deleted")
      .rollback();
    expect(response).toMatchSnapshot();
  });
});
