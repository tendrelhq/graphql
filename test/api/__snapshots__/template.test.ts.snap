// Bun Snapshot v1, https://goo.gl/fbAQLP

exports[`/api/template read 1`] = `
[
  {
    "display_name": {
      "value": "Location",
    },
    "fields": [
      {
        "display_name": {
          "value": "locationname",
        },
        "id": "53350733-73c3-4f6e-8dee-a3581ac1b8a4",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
      {
        "display_name": {
          "value": "locationdisplayname",
        },
        "id": "7bba0fd0-19f4-4984-b8e2-431a5f6c70d0",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
      {
        "display_name": {
          "value": "locationtimezone",
        },
        "id": "2a69677c-b23a-4407-b2ae-1905f0640579",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
      {
        "display_name": {
          "value": "locationlatitude",
        },
        "id": "db4dfca7-2a2c-45d1-a4ec-f749c48d5ddf",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "950dcdeb-e5e7-4e71-a532-78f4a02978d0",
        },
      },
      {
        "display_name": {
          "value": "locationlongitude",
        },
        "id": "37e49e9e-9700-432f-a384-2e68f2279b50",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "950dcdeb-e5e7-4e71-a532-78f4a02978d0",
        },
      },
      {
        "display_name": {
          "value": "locationradius",
        },
        "id": "421b26da-1529-4951-a0ef-9dfb0d18e413",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "950dcdeb-e5e7-4e71-a532-78f4a02978d0",
        },
      },
    ],
    "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
  },
  {
    "display_name": {
      "value": "Customer Tag",
    },
    "fields": [
      {
        "display_name": {
          "value": "custagdisplayname",
        },
        "id": "1b29e7b0-0800-4366-b79e-424dd9bafa71",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
      {
        "display_name": {
          "value": "custagname",
        },
        "id": "7530f263-324e-4ca1-8c08-6b6d31ab2802",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
      {
        "display_name": {
          "value": " custagtype",
        },
        "id": "f6431e52-85af-49d4-9b3b-4eadedb411b3",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
      {
        "display_name": {
          "value": "custagabbreviationentityuuid",
        },
        "id": "d26573c1-c058-48b9-9134-9a925d2ed6da",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "2de2bbde-6319-4886-a58d-bf9d369fc677",
        },
      },
    ],
    "id": "30a317b8-6a56-45b4-8480-f9b58e099c77",
  },
  {
    "display_name": {
      "value": "entitytemplate2025-03-01 16:57:00.851421+00",
    },
    "fields": [],
    "id": "3c04bedf-bcd8-40de-ae35-3a650146f7d7",
  },
  {
    "display_name": {
      "value": "Geolocation",
    },
    "fields": [
      {
        "display_name": {
          "value": "displayname",
        },
        "id": "10b3fb5e-4179-4976-80ec-cb7895d654c4",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
    ],
    "id": "3fa65d6c-f2e0-467d-81a8-10a7e2553bd6",
  },
  {
    "display_name": {
      "value": "entity2025-02-25 21:33:13.332067+00",
    },
    "fields": [],
    "id": "651547d7-825d-408f-a24e-d57362c5e5ac",
  },
  {
    "display_name": {
      "value": "entity2025-02-25 22:18:34.026278+00",
    },
    "fields": [],
    "id": "7a41274c-7829-4ad5-a0ea-e5d42d7b1092",
  },
  {
    "display_name": {
      "value": "entity2025-02-25 22:21:47.409392+00",
    },
    "fields": [],
    "id": "87220492-55cc-467f-85c8-385431806b97",
  },
  {
    "display_name": {
      "value": "Worker",
    },
    "fields": [
      {
        "display_name": {
          "value": "displayname",
        },
        "id": "25652077-06ec-443f-903b-7804a002d9c1",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
    ],
    "id": "957df2f9-051f-4af5-95ee-ea3760fbb83b",
  },
  {
    "display_name": {
      "value": "Customer",
    },
    "fields": [
      {
        "display_name": {
          "value": "customerdisplayname",
        },
        "id": "d15bb9c2-0601-4e4f-9009-c791a40be191",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
      {
        "display_name": {
          "value": "customername",
        },
        "id": "ba834ca4-84cd-49a5-81a7-5676d1df3fbd",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
      {
        "display_name": {
          "value": "customertypeuuid",
        },
        "id": "dfdf9b85-27d2-4d89-8758-5607a2c7cc19",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "2de2bbde-6319-4886-a58d-bf9d369fc677",
        },
      },
      {
        "display_name": {
          "value": "customerlanguagetypeentityuuid",
        },
        "id": "c51fbd4a-dbf5-40a2-892c-edaf81bee4ad",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "2de2bbde-6319-4886-a58d-bf9d369fc677",
        },
      },
    ],
    "id": "b124da10-be8a-4d32-9f68-7f4e6e8b24e9",
  },
  {
    "display_name": {
      "value": "entity2025-02-28 16:34:46.171823+00",
    },
    "fields": [
      {
        "display_name": {
          "value": "fieldname2025-02-28 16:58:35.48333+00",
        },
        "id": "7ec63e60-f3cb-4418-943a-7d239d445295",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "2de2bbde-6319-4886-a58d-bf9d369fc677",
        },
      },
      {
        "display_name": {
          "value": "fieldname2025-02-28 17:11:25.199597+00",
        },
        "id": "2b54f7d5-e051-4bf9-b894-54be4d67a182",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "2de2bbde-6319-4886-a58d-bf9d369fc677",
        },
      },
      {
        "display_name": {
          "value": "fieldname2025-02-28 17:40:31.076587+00",
        },
        "id": "2643d95c-65cf-4e3d-bffa-5e7fbc45f74a",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "2de2bbde-6319-4886-a58d-bf9d369fc677",
        },
      },
      {
        "display_name": {
          "value": "fieldname2025-02-28 17:41:24.634851+00",
        },
        "id": "0ae109a5-9ea9-4d60-aa63-dee867901bed",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "2de2bbde-6319-4886-a58d-bf9d369fc677",
        },
      },
      {
        "display_name": {
          "value": "fieldname2025-03-02 01:41:01.075416+00",
        },
        "id": "c77db174-7b16-4f47-b138-b56766375449",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "2de2bbde-6319-4886-a58d-bf9d369fc677",
        },
      },
    ],
    "id": "cad6913c-071e-4822-9cf6-58cd2d5ff3fd",
  },
  {
    "display_name": {
      "value": "Variety",
    },
    "fields": [
      {
        "display_name": {
          "value": "displayname",
        },
        "id": "47ebdd86-1f13-4a95-8315-f1559d9f8471",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
    ],
    "id": "cfc9cb1b-2121-4b11-9a4f-123b447a01fd",
  },
  {
    "display_name": {
      "value": "System Tag",
    },
    "fields": [
      {
        "display_name": {
          "value": "systagdisplayname",
        },
        "id": "cf94ce9c-edd3-4c7b-8128-ab598fc9710a",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
      {
        "display_name": {
          "value": "systagname",
        },
        "id": "d0a7d4f2-f06b-48ba-ba6d-c7760e9e367e",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
      {
        "display_name": {
          "value": "systagtype",
        },
        "id": "62a0f163-0f7a-40ae-a345-a1258700d755",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
      {
        "display_name": {
          "value": "systagabbreviationentityuuid",
        },
        "id": "29b180af-e230-4e6c-9f56-42bcd0c03ce5",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "2de2bbde-6319-4886-a58d-bf9d369fc677",
        },
      },
    ],
    "id": "def05966-06b2-483e-8988-d0f898e45e6c",
  },
  {
    "display_name": {
      "value": "entity2025-02-25 21:45:37.133743+00",
    },
    "fields": [],
    "id": "e62e67f3-9bd6-4be8-b379-e22d61b51c91",
  },
  {
    "display_name": {
      "value": "Resource",
    },
    "fields": [
      {
        "display_name": {
          "value": "displayname",
        },
        "id": "8a3e0862-173d-4b12-80b0-f21dceefbe0d",
        "type": {
          "display_name": {
            "value": "System Tag",
          },
          "id": "82292605-5671-488c-9f8c-e821d5d7eb4e",
        },
      },
    ],
    "id": "ea6b76fa-7d8c-45bc-ab72-fd106af2a5a5",
  },
  {
    "display_name": {
      "value": "entity2025-02-25 21:35:35.143757+00",
    },
    "fields": [],
    "id": "f7f39a93-9c82-4128-9902-b23a0355699d",
  },
]
`;

exports[`/api/template delete 1`] = `
{
  "count": null,
  "data": [
    {
      "_deleted": true,
      "id": "0b9f3142-e7ed-4f78-8504-ccd2eb505075",
    },
  ],
  "error": null,
  "status": 200,
  "statusText": "OK",
}
`;
